File: packages/shared-kernel/src/contracts/ai/model-contract.ts
export type GestureLabel = {
  id: string;
  name: string;
  actionType:
    | "open-app"
    | "volume-down"
    | "volume-up"
    | "mute"
    | "keyboard-shortcut"
    | "mouse-click";
  actionTarget: string;
};

export type GestureSampleReference = {
  id: string;
  gestureId: string;
  capturedAt: string;
  filePath: string;
  width?: number;
  height?: number;
  handDetected?: boolean;
  handDetectionConfidence?: number | null;
  handLandmarkCount?: number | null;
  source: "desktop-camera";
};

export type TrainingDataset = {
  id: string;
  name: string;
  labels: GestureLabel[];
  sampleCount: number;
  createdAt: string;
  updatedAt: string;
};

export type TrainingJobStatus =
  | "queued"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

export type TrainingEvaluationMetrics = {
  accuracy: number | null;
  confusionMatrix: Record<string, Record<string, number>>;
  perGestureAccuracy: Record<string, number>;
  sampleCount: number;
};

export type TrainingQualityMetrics = {
  training?: TrainingEvaluationMetrics;
  validation?: TrainingEvaluationMetrics;
};

export type TrainingJob = {
  id: string;
  datasetId: string;
  modelFamily: "gesture-recognition";
  status: TrainingJobStatus;
  progress: number;
  startedAt?: string;
  completedAt?: string;
  modelArtifactPath?: string;
  errorMessage?: string;
  metrics?: TrainingQualityMetrics;
};

export type ModelStatus = {
  modelId: string;
  modelFamily: "gesture-recognition" | "continuous-authentication";
  status: "not-trained" | "loading" | "ready" | "degraded" | "error";
  version?: string;
  accuracy?: number;
  latencyMs?: number;
  loadedAt?: string;
  errorMessage?: string;
};

export type InferenceFrameReference = {
  capturedAt: string;
  frameId: string;
  filePath?: string;
};

export type GesturePrediction = {
  gestureId: string;
  label: string;
  confidence: number;
};

export type InferenceResult = {
  id: string;
  frameId: string;
  modelId: string;
  predictions: GesturePrediction[];
  bestPrediction?: GesturePrediction;
  inferenceTimeMs: number;
  createdAt: string;
};

export type HandPresenceResult = {
  frameId: string;
  handDetected: boolean;
  confidence?: number | null;
  landmarkCount?: number | null;
  boundingBox?: {
    x: number;
    y: number;
    width: number;
    height: number;
  } | null;
  reason?: string | null;
};

export type PredictionEvent = {
  id: string;
  resultId: string;
  gestureId?: string;
  label: string;
  confidence: number;
  accepted: boolean;
  createdAt: string;
};

export type CreateTrainingDatasetRequest = {
  name: string;
  labels: GestureLabel[];
  samples: GestureSampleReference[];
};

export type TrainGestureModelRequest = {
  datasetId: string;
  modelConfig?: {
    epochs?: number;
    batchSize?: number;
    learningRate?: number;
  };
};

export type RunGestureInferenceRequest = {
  frame: InferenceFrameReference;
  modelId?: string;
  minConfidence?: number;
};

export type DetectHandPresenceRequest = {
  frame: InferenceFrameReference;
};

export interface AiModelServiceContract {
  createTrainingDataset(request: CreateTrainingDatasetRequest): Promise<TrainingDataset>;
  detectHandPresence(request: DetectHandPresenceRequest): Promise<HandPresenceResult>;
  getModelStatus(modelId?: string): Promise<ModelStatus>;
  getTrainingJob(jobId: string): Promise<TrainingJob>;
  listDatasets(): Promise<TrainingDataset[]>;
  runGestureInference(request: RunGestureInferenceRequest): Promise<InferenceResult>;
  trainGestureModel(request: TrainGestureModelRequest): Promise<TrainingJob>;
}

File: services/ai-models/src/interfaces/api/schemas/model_contract.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


GestureActionType = Literal[
    "open-app",
    "volume-down",
    "volume-up",
    "mute",
    "keyboard-shortcut",
    "mouse-click",
]


class GestureLabelSchema(BaseModel):
    id: str
    name: str
    actionType: GestureActionType
    actionTarget: str


class GestureSampleReferenceSchema(BaseModel):
    id: str
    gestureId: str
    capturedAt: str
    filePath: str
    source: Literal["desktop-camera"]
    width: int | None = None
    height: int | None = None


class TrainingDatasetSchema(BaseModel):
    id: str
    name: str
    labels: list[GestureLabelSchema]
    sampleCount: int
    createdAt: str
    updatedAt: str


class ModelConfigSchema(BaseModel):
    epochs: int | None = None
    batchSize: int | None = None
    learningRate: float | None = None


class CreateTrainingDatasetRequestSchema(BaseModel):
    name: str
    labels: list[GestureLabelSchema]
    samples: list[GestureSampleReferenceSchema]


class TrainGestureModelRequestSchema(BaseModel):
    datasetId: str
    modelConfig: ModelConfigSchema | None = None


class TrainingJobSchema(BaseModel):
    id: str
    datasetId: str
    modelFamily: Literal["gesture-recognition"] = "gesture-recognition"
    status: Literal["queued", "running", "completed", "failed", "cancelled"]
    progress: float = Field(ge=0, le=1)
    startedAt: str | None = None
    completedAt: str | None = None
    modelArtifactPath: str | None = None
    errorMessage: str | None = None


class ModelStatusSchema(BaseModel):
    modelId: str
    modelFamily: Literal["gesture-recognition", "continuous-authentication"]
    status: Literal["not-trained", "loading", "ready", "degraded", "error"]
    version: str | None = None
    accuracy: float | None = None
    latencyMs: float | None = None
    loadedAt: str | None = None
    errorMessage: str | None = None


class InferenceFrameReferenceSchema(BaseModel):
    capturedAt: str
    frameId: str
    filePath: str | None = None


class GesturePredictionSchema(BaseModel):
    gestureId: str
    label: str
    confidence: float = Field(ge=0, le=1)


class RunGestureInferenceRequestSchema(BaseModel):
    frame: InferenceFrameReferenceSchema
    modelId: str | None = None
    minConfidence: float | None = Field(default=None, ge=0, le=1)


class DetectHandPresenceRequestSchema(BaseModel):
    frame: InferenceFrameReferenceSchema


class InferenceResultSchema(BaseModel):
    id: str
    frameId: str
    modelId: str
    predictions: list[GesturePredictionSchema]
    bestPrediction: GesturePredictionSchema | None = None
    inferenceTimeMs: float
    createdAt: str


class HandPresenceResultSchema(BaseModel):
    frameId: str
    handDetected: bool
    reason: str | None = None
